codec-http2/src/main/java/io/netty/handler/codec/http2/Http2MultiplexCodec.java

```diff
@@ -137,145 +242,101 @@ public void channelRead(ChannelHandlerContext ctx, Object msg) throws Exception
 
-    private void onStreamClosed0(Http2StreamChannel childChannel) {
-        assert childChannel.eventLoop().inEventLoop();
+    private void onStreamActive(Http2FrameStream stream) {
+        DefaultHttp2StreamChannel childChannel = newStreamChannel(stream);
 
-        childChannel.onStreamClosedFired = true;
-        childChannel.fireChildRead(CLOSE_MESSAGE);
+        childChannel.pipeline().addLast(inboundStreamHandler);
+
+        ChannelFuture future = ctx.channel().eventLoop().register(childChannel);
+        future.addListener(CHILD_CHANNEL_REGISTRATION_LISTENER);
     }
 
-    void flushFromStreamChannel() {
-        EventExecutor executor = ctx.executor();
-        if (executor.inEventLoop()) {
-            flush(ctx);
-        } else {
-            Runnable task = flushTask;
-            if (task == null) {
-                task = flushTask = new Runnable() {
-                    @Override
-                    public void run() {
-                        flush(ctx);
-                    }
-                };
-            }
-            executor.execute(task);
-        }
+    // TODO: This is most likely not the best way to expose this, need to think more about it.
+    final Http2StreamChannel newOutboundStream() {
+        return newStreamChannel(newStream());
     }
 
-    void writeFromStreamChannel(Object msg, boolean flush) {
-        writeFromStreamChannel(msg, ctx.newPromise(), flush);
+    private DefaultHttp2StreamChannel newStreamChannel(Http2FrameStream stream) {
+        DefaultHttp2StreamChannel childChannel = new DefaultHttp2StreamChannel(stream);
+        // TODO: Maybe assert that this return null ?
+        channels.put(stream, childChannel);
+        return childChannel;
     }
 
-    void writeFromStreamChannel(final Object msg, final ChannelPromise promise, final boolean flush) {
-        EventExecutor executor = ctx.executor();
-        if (executor.inEventLoop()) {
-            writeFromStreamChannel0(msg, flush, promise);
-        } else {
+    @Override
+    public void exceptionCaught(ChannelHandlerContext ctx, Throwable cause) throws Exception {
+        if (cause instanceof Http2FrameStreamException) {
+            Http2FrameStreamException streamException = (Http2FrameStreamException) cause;
+            Http2FrameStream stream = streamException.stream();
+            DefaultHttp2StreamChannel childChannel = channels.get(stream);
+
+            if (childChannel == null)  {
+                // TODO: Should we log this ?
+                return;
+            }
             try {
-                executor.execute(new Runnable() {
-                    @Override
-                    public void run() {
-                        writeFromStreamChannel0(msg, flush, promise);
-                    }
-                });
-            } catch (Throwable cause) {
-                promise.setFailure(cause);
+                childChannel.pipeline().fireExceptionCaught(streamException.getCause());
+            } finally {
```

Comment:
![INFO](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-info.png) Complete the task associated to this TODO comment. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS1135)